api/query/search.go

```diff
@@ -99,51 +110,26 @@ func (sh structuredSearchHandler) ServeHTTP(w http.ResponseWriter, r *http.Reque
 			val, _ := shared.ParseBooleanParam(q, param)
 			isSimpleQ = isSimpleQ && (val == nil || !*val)
 		}
-	}
-
-	if !isSimpleQ {
-		ctx := sh.api.Context()
-		hostname := sh.api.GetServiceHostname("searchcache")
-		// TODO: This will not work when hostname is localhost (http scheme needed).
-		fwdURL, _ := url.Parse(fmt.Sprintf("https://%s/api/search/cache", hostname))
-		fwdURL.RawQuery = r.URL.RawQuery
 
-		logger := shared.GetLogger(ctx)
-		logger.Infof("Forwarding structured search request to %s: %s", hostname, string(data))
-
-		client := sh.api.GetHTTPClientWithTimeout(time.Second * 15)
-		req, err := http.NewRequest("POST", fwdURL.String(), bytes.NewBuffer(data))
-		if err != nil {
-			logger.Errorf("Failed to create request to POST %s: %v", fwdURL.String(), err)
-			http.Error(w, "Error connecting to search API cache", http.StatusInternalServerError)
-			return
+		// Check old summary files. If any can't be found,
+		// use the searchcache to aggregate the runs.
+		for _, id := range rq.RunIDs {
+			q.Add("run_id", strconv.FormatInt(id, 10))
 		}
-		req.Header.Add("Content-Type", "application/json")
+		q.Add("q", url.QueryEscape(simpleQ.Pattern))
 
-		resp, err := client.Do(req)
+		r2 := r.Clone(r.Context())
+		r2.Method = "GET"
+		r2.URL.RawQuery = q.Encode()
+		isSimpleQ, err = sh.validateSummaryVersions(r2.URL.Query())
 		if err != nil {
-			logger.Errorf("Error connecting to search API cache: %v", err)
-			http.Error(w, "Error connecting to search API cache", http.StatusInternalServerError)
-			return
-		}
-
-		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-			msg := fmt.Sprintf("Error from request: POST %s: STATUS %d", fwdURL.String(), resp.StatusCode)
-			errBody, err2 := ioutil.ReadAll(resp.Body)
-			if err2 == nil {
-				msg = fmt.Sprintf("%s: %s", msg, string(errBody))
-				resp.Body = ioutil.NopCloser(bytes.NewBuffer(errBody))
-			}
-			logger.Errorf(msg)
+			logger.Debugf("Error checking version files: %v", err)
 		}
+	}
 
-		defer resp.Body.Close()
-		w.WriteHeader(resp.StatusCode)
-		_, err = io.Copy(w, resp.Body)
-		if err != nil {
-			logger.Errorf("Error forwarding response payload from search cache: %v", err)
-		}
-		return
+	// Use searchcache for a complex query or if old summary files exist.
+	if !isSimpleQ {
+		sh.useSearchcache(w, r, data, logger)
```

Comment:
This function writes to the response writer. You should return the appropriate error / success and write it to the response writer at this level.

Reply:
Good idea! I'll refactor to do so.